lib/protobuf/json/decode.ex

```diff
@@ -0,0 +1,207 @@
+defmodule Protobuf.JSON.Decode do
+  @moduledoc false
+
+  import Bitwise, only: [bsl: 2]
+
+  alias Protobuf.JSON.Utils
+
+  @compile {:inline,
+            field_value: 2,
+            decode_map: 2,
+            decode_repeated: 2,
+            decode_integer: 1,
+            decode_float: 1,
+            parse_float: 1,
+            decode_bytes: 1,
+            decode_key: 3,
+            parse_key: 2}
+
+  @int32_range -bsl(1, 31)..(bsl(1, 31) - 1)
+  @int64_range -bsl(1, 63)..(bsl(1, 63) - 1)
+  @uint32_range 0..(bsl(1, 32) - 1)
+  @uint64_range 0..(bsl(1, 64) - 1)
+
+  @int_ranges %{
+    int32: @int32_range,
+    int64: @int64_range,
+    sint32: @int32_range,
+    sint64: @int64_range,
+    sfixed32: @int32_range,
+    sfixed64: @int64_range,
+    fixed32: @int32_range,
+    fixed64: @int64_range,
+    uint32: @uint32_range,
+    uint64: @uint64_range
+  }
+
+  @int_types Map.keys(@int_ranges)
+
+  @float_types [:float, :double]
+
+  def from_json_data(data, module) when is_map(data) do
+    message_props = Utils.message_props(module)
+    regular = decode_regular_fields(data, message_props)
+    oneofs = decode_oneof_fields(data, message_props)
+
+    module.__default_struct__()
+    |> struct(regular)
+    |> struct(oneofs)
+  end
+
+  def from_json_data(data, _module), do: throw({:bad_message, data})
```

Comment:
We'd have to do this here too if we go with the suggestion above:

```suggestion
  def from_json_data(data, module) when is_atom(module), do: throw({:bad_message, data})
```